src/python/pants/util/meta.py

```diff
@@ -77,12 +81,17 @@ def classproperty(func: Union[classmethod, staticmethod, Callable]) -> ClassProp
   doc = func.__doc__
 
   if not isinstance(func, (classmethod, staticmethod)):
-    func = classmethod(func)
+    # MyPy complains about converting a Callable -> classmethod. We use a Callable in the first
+    # place because there is no typing.classmethod, i.e. a type that takes generic arguments, and
+    # we need to use TypeVars for the call sites of this decorator to work properly.
+    func = classmethod(func)  # type: ignore[assignment]
 
-  return ClassPropertyDescriptor(func, doc)
+  # If we properly annotated this function as returning a ClassPropertyDescriptor, then MyPy would
+  # no longer work correctly at call sites for this decorator.
```

Comment:
What if we defined `P = TypeVar('P', bound=ClassPropertyDescriptor)`?

Reply:
This wouldn't work as intended because then you could no longer type something like:

```python
@classproperty
def foo() -> str:
  return "hi"
```

`str` is not a `ClassPropertyDescriptor`.